monkey/infection_monkey/network/relay/relay_user_handler.py

```diff
@@ -0,0 +1,70 @@
+from dataclasses import dataclass
+from ipaddress import IPv4Address
+from threading import Lock
+from time import time
+from typing import Dict
+
+DEFAULT_NEW_CLIENT_TIMEOUT = 3  # Wait up to 3 seconds for potential new clients to connect
+
+
+@dataclass
+class RelayUser:
+    address: IPv4Address
+    last_update_time: float
+
+
+class RelayUserHandler:
+    def __init__(self, new_client_timeout: float = DEFAULT_NEW_CLIENT_TIMEOUT):
+        self._new_client_timeout = new_client_timeout
+        self._relay_users: Dict[IPv4Address, RelayUser] = {}
+        self._potential_users: Dict[IPv4Address, RelayUser] = {}
+
+        self._lock = Lock()
+
+    def add_relay_user(self, user_address: IPv4Address):
+        """
+        Handle new user connection.
+
+        :param source_socket: A source socket
+        :param user_address: An address defining RelayUser which will be added to the relay
+        """
+
+        with self._lock:
+            if user_address in self._potential_users:
+                del self._potential_users[user_address]
```

Comment:
Use [`common.utils.code_utils.del_key()`](https://github.com/guardicore/monkey/blob/de5da88c22827ca67980646710e609947cb8bce3/monkey/common/utils/code_utils.py#L37-L50) 

https://github.com/guardicore/monkey/blob/de5da88c22827ca67980646710e609947cb8bce3/monkey/common/utils/code_utils.py#L37-L50

Reply:
We are checking that the key exists before calling `del`. Are you saying replace the entire `if` statement with `del_key()`? Note also that `self._potential_users.pop(user_address, None)` would delete the key without raising an error. I'm not sure which is the preferred method

Reply:
I'm saying replace the entire `if`. Maybe `pop()` is a better solution than `del_key()`. I'll leave that up to you.

Reply:
I'll use `del_key()`. We can always change the implementation of `del_key()` if we decide using `pop()` is better than handling the exception